src/storage/sftp/storage.c

```diff
@@ -37,6 +38,199 @@ struct StorageSftp
+                    Storage *sshStorage =
+                        storagePosixNewP(
+                            strNewFmt("%s%s", strZ(userHome()), "/.ssh"), .modeFile = 0644, .modePath = 0700, .write = true);
+
+                    if (!storageExistsP(sshStorage, strNewFmt("%s", "known_hosts")))
+                        storagePutP(storageNewWriteP(sshStorage, strNewFmt("%s", "known_hosts")), NULL);
+
+                    // Try to load the user's known_hosts file entries into the collection again
+                    rc = libssh2_knownhost_readfile(userKnownHostsList, userKnownHostsFile, LIBSSH2_KNOWNHOST_FILE_OPENSSH);
+                }
+            }
+
+            // If the user's known_hosts file was read successfully, add the host to the collection and rewrite the file.
+            if (rc >= 0)
+            {
+                // Check for a supported known host key type
+                const int knownHostKeyType = storageSftpKnownHostKeyType(hostKeyType);
+
+                if (knownHostKeyType != 0)
+                {
+                    // Add host to the internal list
+                    if (libssh2_knownhost_addc(userKnownHostsList, strZ(host), NULL, hostKey, hostKeyLen,
+                            strZ(strNewZ("Generated from " PROJECT_NAME)), strlen(strZ(strNewZ("Generated from " PROJECT_NAME))),
+                            LIBSSH2_KNOWNHOST_TYPE_PLAIN | LIBSSH2_KNOWNHOST_KEYENC_RAW | knownHostKeyType, NULL) == 0)
+                    {
+                        // Rewrite the updated known_hosts file
+                        rc = libssh2_knownhost_writefile(userKnownHostsList, userKnownHostsFile, LIBSSH2_KNOWNHOST_FILE_OPENSSH);
+
+                        if (rc != 0)
+                            LOG_WARN_FMT(PROJECT_NAME " unable to write '%s' for update", userKnownHostsFile);
+                        else
+                            LOG_WARN_FMT(PROJECT_NAME " added new host '%s' to '%s'", strZ(host), userKnownHostsFile);
+                    }
+                    else
+                        LOG_WARN_FMT(PROJECT_NAME " failed to add '%s' to known_hosts internal list", strZ(host));
+                }
+                else
+                    LOG_WARN_FMT("unsupported key type [%d], unable to update knownhosts for '%s'", hostKeyType, strZ(host));
+            }
+            else
+            {
+                // On readfile failure warn that we're unable to update the user's known_hosts file
+                rc = libssh2_session_last_error(this->session, &libSsh2ErrMsg, &libSsh2ErrMsgLen, 0);
+
+                LOG_WARN_FMT(
+                    "libssh2 unable to read '%s' for update: libssh2 errno [%d] %s\n"
+                    "HINT: does '%s' exist with proper permissions?", userKnownHostsFile, rc, libSsh2ErrMsg, userKnownHostsFile);
+            }
+        }
+
+        // Free the user's known hosts list
+        if (userKnownHostsList)
+            libssh2_knownhost_free(userKnownHostsList);
```

Comment:
Seems like all callers init their own and will handle free'ing it too, so maybe we should just use that one instead?  See earlier comment.

Reply:
The initial init will load up the entries in either the default hosts file list or the file(s) provided by the repo-sftp-known-host values. The storageSftpUpdateKnownHostsFile() function needs to operate on only the user's known_hosts file. If we use the initial collection then we'd write out all entries from all files to the user's known_hosts when we add a new host entry.